File: mcp-server/packages/context-extractor-csharp/src/Models/ObjectNode.cs
using System.Text;
using System.Text.Json;

namespace ContextExtractor.Models;

/// <summary>
/// Represents a node in the object hierarchy tree.
/// Contains information about an object's class, methods, and child dependencies.
/// </summary>
public class ObjectNode
{
    public string Name { get; set; }
    public string ClassName { get; set; }
    public List<MethodInfo> Methods { get; set; } = new();
    public List<ObjectNode> Children { get; set; } = new();
    public int Depth { get; set; }

    public ObjectNode(string name, string className, int depth = 0)
    {
        Name = name;
        ClassName = className;
        Depth = depth;
    }

    /// <summary>
    /// Count total methods in this node and all children.
    /// </summary>
    public int CountTotalMethods()
    {
        int count = Methods.Count;
        foreach (var child in Children)
        {
            count += child.CountTotalMethods();
        }
        return count;
    }

    /// <summary>
    /// Count total dependencies (children) recursively.
    /// </summary>
    public int CountTotalDependencies()
    {
        int count = Children.Count;
        foreach (var child in Children)
        {
            count += child.CountTotalDependencies();
        }
        return count;
    }

    /// <summary>
    /// Convert to JSON representation.
    /// </summary>
    public string ToJson(bool indented = true)
    {
        var options = new JsonSerializerOptions
        {
            WriteIndented = indented,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };
        return JsonSerializer.Serialize(ToDictionary(), options);
    }

    public Dictionary<string, object?> ToDictionary()
    {
        return new Dictionary<string, object?>
        {
            ["name"] = Name,
            ["class"] = ClassName,
            ["depth"] = Depth,
            ["methods"] = Methods.Select(m => m.ToDictionary()).ToList(),
            ["children"] = Children.Select(c => c.ToDictionary()).ToList()
        };
    }

    /// <summary>
    /// Convert to AI-optimized text representation.
    /// </summary>
    public string ToText(bool includeDetails = true)
    {
        var sb = new StringBuilder();
        BuildTextTree(sb, includeDetails, "");
        return sb.ToString();
    }

    private void BuildTextTree(StringBuilder sb, bool includeDetails, string prefix)
    {
        // Add header
        if (Depth == 0)
        {
            sb.AppendLine($"# {ClassName} Component Structure");
            sb.AppendLine();
            sb.AppendLine($"Root object: {Name} -> {ClassName}");
        }
        else
        {
            var arrow = Children.Count == 0 ? "└──" : "├──";
            sb.AppendLine($"{prefix}{arrow} {Name}: {ClassName}");
        }

        // Add methods
        if (Methods.Count > 0)
        {
            var methodPrefix = Depth > 0 ? prefix + "    " : "";
            if (Depth == 0)
            {
                sb.AppendLine();
                sb.AppendLine("## Methods");
            }
            foreach (var method in Methods)
            {
                sb.AppendLine($"{methodPrefix}  → .{method.GetSignature()}");
                if (includeDetails && !string.IsNullOrEmpty(method.Docstring))
                {
                    var firstLine = method.Docstring.Split('\n')[0].Trim();
                    sb.AppendLine($"{methodPrefix}      # {firstLine}");
                }
            }
        }

        // Add children
        if (Children.Count > 0)
        {
            if (Depth == 0)
            {
                sb.AppendLine();
                sb.AppendLine("## Dependencies");
            }
            for (int i = 0; i < Children.Count; i++)
            {
                var isLast = i == Children.Count - 1;
                var childPrefix = Depth > 0 ? prefix + (isLast ? "    " : "│   ") : "";
                Children[i].BuildTextTree(sb, includeDetails, childPrefix);
            }
        }

        // Add summary at root level
        if (Depth == 0)
        {
            sb.AppendLine();
            sb.AppendLine("## Summary");
            sb.AppendLine($"- Total methods: {CountTotalMethods()}");
            sb.AppendLine($"- Total dependencies: {CountTotalDependencies()}");
        }
    }
}

File: mcp-server/packages/context-extractor-java/src/main/java/io/contextextractor/ObjectNode.java
package io.contextextractor;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Represents a node in the object hierarchy tree.
 * Contains information about an object's class, methods, and child dependencies.
 */
public class ObjectNode {
    private final String name;
    private final String className;
    private final List<MethodInfo> methods;
    private final List<ObjectNode> children;
    private final int depth;

    private ObjectNode(Builder builder) {
        this.name = builder.name;
        this.className = builder.className;
        this.methods = builder.methods;
        this.children = builder.children;
        this.depth = builder.depth;
    }

    public String getName() {
        return name;
    }

    public String getClassName() {
        return className;
    }

    public List<MethodInfo> getMethods() {
        return new ArrayList<>(methods);
    }

    public List<ObjectNode> getChildren() {
        return new ArrayList<>(children);
    }

    public int getDepth() {
        return depth;
    }

    /**
     * Count total methods in this node and all children.
     */
    public int countTotalMethods() {
        int count = methods.size();
        for (ObjectNode child : children) {
            count += child.countTotalMethods();
        }
        return count;
    }

    /**
     * Count total dependencies (children) recursively.
     */
    public int countTotalDependencies() {
        int count = children.size();
        for (ObjectNode child : children) {
            count += child.countTotalDependencies();
        }
        return count;
    }

    /**
     * Convert to JSON representation.
     */
    public String toJson() {
        return toJson(0);
    }

    private String toJson(int indent) {
        String spaces = " ".repeat(indent);
        String innerSpaces = " ".repeat(indent + 2);
        StringBuilder sb = new StringBuilder();
        
        sb.append(spaces).append("{\n");
        sb.append(innerSpaces).append("\"name\": \"").append(escapeJson(name)).append("\",\n");
        sb.append(innerSpaces).append("\"class\": \"").append(escapeJson(className)).append("\",\n");
        sb.append(innerSpaces).append("\"depth\": ").append(depth).append(",\n");
        
        // Methods
        sb.append(innerSpaces).append("\"methods\": [\n");
        for (int i = 0; i < methods.size(); i++) {
            sb.append(" ".repeat(indent + 4)).append(methods.get(i).toJson());
            if (i < methods.size() - 1) sb.append(",");
            sb.append("\n");
        }
        sb.append(innerSpaces).append("],\n");
        
        // Children
        sb.append(innerSpaces).append("\"children\": [\n");
        for (int i = 0; i < children.size(); i++) {
            sb.append(children.get(i).toJson(indent + 4));
            if (i < children.size() - 1) sb.append(",");
            sb.append("\n");
        }
        sb.append(innerSpaces).append("]\n");
        
        sb.append(spaces).append("}");
        return sb.toString();
    }

    /**
     * Convert to AI-optimized text representation.
     */
    public String toText() {
        return toText(true);
    }

    public String toText(boolean includeDetails) {
        StringBuilder sb = new StringBuilder();
        buildTextTree(sb, includeDetails, "");
        return sb.toString();
    }

    private void buildTextTree(StringBuilder sb, boolean includeDetails, String prefix) {
        // Add header
        if (depth == 0) {
            sb.append("# ").append(className).append(" Component Structure\n\n");
            sb.append("Root object: ").append(name).append(" -> ").append(className).append("\n");
        } else {
            String arrow = children.isEmpty() ? "└──" : "├──";
            sb.append(prefix).append(arrow).append(" ").append(name).append(": ").append(className).append("\n");
        }

        // Add methods
        if (!methods.isEmpty()) {
            String methodPrefix = depth > 0 ? prefix + "    " : "";
            if (depth == 0) {
                sb.append("\n## Methods\n");
            }
            for (MethodInfo method : methods) {
                sb.append(methodPrefix).append("  → .").append(method.getSignature()).append("\n");
                if (includeDetails && method.getDocstring() != null && !method.getDocstring().isEmpty()) {
                    String firstLine = method.getDocstring().split("\n")[0].trim();
                    sb.append(methodPrefix).append("      # ").append(firstLine).append("\n");
                }
            }
        }

        // Add children
        if (!children.isEmpty()) {
            if (depth == 0) {
                sb.append("\n## Dependencies\n");
            }
            for (int i = 0; i < children.size(); i++) {
                boolean isLast = i == children.size() - 1;
                String childPrefix = depth > 0 ? prefix + (isLast ? "    " : "│   ") : "";
                children.get(i).buildTextTree(sb, includeDetails, childPrefix);
            }
        }

        // Add summary at root level
        if (depth == 0) {
            sb.append("\n## Summary\n");
            sb.append("- Total methods: ").append(countTotalMethods()).append("\n");
            sb.append("- Total dependencies: ").append(countTotalDependencies()).append("\n");
        }
    }

    private String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    public static Builder builder(String name, String className) {
        return new Builder(name, className);
    }

    public static class Builder {
        private final String name;
        private final String className;
        private List<MethodInfo> methods = new ArrayList<>();
        private List<ObjectNode> children = new ArrayList<>();
        private int depth = 0;

        public Builder(String name, String className) {
            this.name = name;
            this.className = className;
        }

        public Builder methods(List<MethodInfo> methods) {
            this.methods = new ArrayList<>(methods);
            return this;
        }

        public Builder addMethod(MethodInfo method) {
            this.methods.add(method);
            return this;
        }

        public Builder children(List<ObjectNode> children) {
            this.children = new ArrayList<>(children);
            return this;
        }

        public Builder addChild(ObjectNode child) {
            this.children.add(child);
            return this;
        }

        public Builder depth(int depth) {
            this.depth = depth;
            return this;
        }

        public ObjectNode build() {
            return new ObjectNode(this);
        }
    }
}
